Changes requested by DWA Group:

krishmaj opened the pull request:
- the component testing history entries now show the test times as well as dates
- added the ability to query a single Wire Tension value on the Wire Tension test results page
- improved the layout of the Wire Tension test results page and plots within, and fixed a bug where the wire numbers were being displayed on half-integer values
- fixed a bug with incorrect links to components from within the component and test type forms
- fixed a small code formatting inconsistency in the edit_component.pug file
- removed the explicit list of bugs from the homepage, and changed to now direct the user to the DB's Github issues pages

krishmaj committed:
Changes requested by DWA Group:
- the component testing history entries now show the test times as well as dates
- added the ability to query a single Wire Tension value on the Wire Tension test results page
- improved the layout of the Wire Tension test results page and plots within, and fixed a bug where the wire numbers were being displayed on half-integer values
- fixed a bug with incorrect links to components from within the component and test type forms
- fixed a small code formatting inconsistency in the edit_component.pug file
- removed the explicit list of bugs from the homepage, and changed to now direct the user to the DB's Github issues pages

micchickenburger commented on pug/home.pug:
```diff
@@ -57,10 +57,13 @@ block content
-      li All component pages - the component history is unavailable (null date error)
+      li Bugs and other errors with existing functionality should be reported at our 
+         |
+         a(href = "https://github.com/DUNE/dunedb/issues") Github Issues page
```
Consider whether adding `target = "_blank"` would be better to avoid navigating away from the page if the user needs to report a bug.

krishmaj replied:
Will do.

micchickenburger commented on pug/test_viewResult.pug:
```diff
@@ -37,6 +37,21 @@ block extrascripts
+      var side = document.getElementById("apaSide").value;
+      var number = document.getElementById("wireNumber").value;
+      
+      var keyName = "wires." + layer + "." + side + "[" + number + "]";
```
Just as an FYI, you can use string interpolation instead of concatenation:
```
const keyName = `wires.${layer}.${side}[${number}]`;
```

krishmaj replied:
I actually tried this exactly as you've written, and it didn't work for some reason.

micchickenburger replied:
If the code gets transpiled, the PUG transpiler might not support ESNext, in which case we might just need to update it.  Natively, however, all modern browsers support this feature.

krishmaj replied:
To be fair, I don't actually know why it wasn't working ... I'm aware that string interpolation should work with my browser, so I knew it wasn't that which was the issue.  I'll give it another go ... might have been something else on the page that was breaking things (I was playing with other things on it at the same time).

micchickenburger commented on pug/test_viewResult.pug:
```diff
@@ -71,7 +86,39 @@ block content
+          .col-3
+      
+        pre #{''}
+        pre #{''}
```
Why are preformatted elements used here?

krishmaj replied:
I've found that it's the best way to just get some spacing between elements and at the bottom of pages ... most of the other methods that I've tried don't consistently work for whatever reason.

micchickenburger replied:
[Broken Windows Theory](https://en.wikipedia.org/wiki/Broken_windows_theory), though appropriately criticized in the psychology and criminology fields where it originated, definitely applies to software engineering.  Hacking things to work achieves the desired outcome but perpetuates more hacks in the future as contributors care less and less about code quality.  And of course code quality is directly tied to maintainability (poor code quality creates technical debt, which manifests as increased cost in time and resources for writing bug-free code because existing code is unclear, undocumented, or buggy.)

So in this case `pre` elements are only intended to be used to render preformatted text.  To use them for any other purpose creates technical debt.  Sometimes there are no obvious solutions for a problem, in which case a hack may be necessary.  But if that's the case the hack must be clearly documented:
* This is the problem we face
* This is the solution we chose and why
* These are the expected inputs
* This is the expected output

If the problem is adding spacing, the solution is to use CSS.  CSS is the solution to every style-related issue.  HTML should *only* markup data (e.g. this is a table of data, this is a header, this is a paragraph, etc.)  Colors, borders, outlines, padding, margins, backgrounds, animations, shapes, etc--these are all within the realm of CSS.

krishmaj replied:
Fair enough.  Now, baring in mind that I've done the exact same thing in every other pug file that I've looked over in the past 2 months, can we leave this here for this commit and I will put in a separate commit where I scrub all of the pre statements in favour of the correct code?  Would like to have some time to try out what works and what doesn't, rather than rushing a change for this commit.

micchickenburger replied:
Sure, I think that's fine.  Creating an issue for this would be a good idea too so others would know there is intention to remove them.  :)

micchickenburger commented on pug/test_viewResult.pug:
```diff
@@ -132,3 +178,6 @@ block content
       a.btn.btn-info(href = '/test/' + testdata._id + '/copyToDraft') Copy Test to New Draft
   
+    pre #{''}
+    pre #{''}
```
Why are preformatted elements used here?

krishmaj replied:
As above.

micchickenburger commented on static/components/ComponentArray.js:
```diff
@@ -58,25 +58,21 @@ class ArrayComponent extends TextFieldComponent{
+        tpl += `</div>`
+        tpl += `<div class='flex-grow-1 p-4 arrayComponentGraph' style='height: 200px; width: 240px;'></div>`;
+        tpl += `<div class='flex-grow-1 p-4 arrayComponentHistogram' style='height: 200px; width: 240px;'></div>`;
+      tpl += `</div>`
```
Consider using multiline string interpolation instead of concatenation for better readability:
```
tpl = `
  <div>Max: ... </div>
  <div>Mean: ... </div>
  ...
  <div>
    <span>Number of Wires...</span>
  </div>
`;
```

krishmaj replied:
Good point - was mainly just trying to keep as close to the original code as I could, but why keep it difficult.

krishmaj replied:
I've changed as much as I could without breaking the formatting ... but I'm sure it can be improved further (I'm not yet so good with combining string interpolation and HTML tags), so I'm happy to make further changes.

micchickenburger replied:
I cleaned up that file a bit more, stylistically.  See how the string interpolation can be accomplished with variables?  In the case of conditionals, ternary expressions can be used.
<img width="823" alt="image" src="https://user-images.githubusercontent.com/4521230/144663838-13ad4180-e113-4926-b57c-59d05be7383a.png">

micchickenburger commented on static/components/ComponentArray.js:
```diff
@@ -237,11 +232,6 @@ class ArrayComponent extends TextFieldComponent{
     console.log("this.value",this.value);
     // debugger;
```
I know this isn't your code but it would be a good idea to remove all commented code and console statements from the code.  If we need to reference deleted code, we can look at the file history in git.  It's just a code hygiene thing.  One big question everyone reading the code will have is "why is this code commented out?"

krishmaj replied:
Is on my todo list actually - exactly as you've said, to remove all unused code and logger/console lines.  Wanted to get more of the DB checked over first before doing that though.

krishmaj replied:
I'll make an issue for this one as well, again so that people know it's coming.

micchickenburger commented on static/components/ComponentUUID_Component.js:
```diff
@@ -126,7 +126,7 @@ class ComponentUUID_Component extends QR_Component{
     // }
     if(this.refs.compUuidInfo && value && value.length==36) {
```
I don't know if you know the answer to this, but `value.length==36` begs a comment for explanation.

krishmaj replied:
No idea to be honest.  Can investigate if you want.

krishmaj replied:
The 36 refers to the total length (number of characters) in the UUID pattern: 8 + 4 + 4 + 4+ 12, plus 4 dashes.

krishmaj committed:
Additional changes based on pull request comments:
- added the blank target to the homepage Issues redirect
- changed to using string interpolation for building the keyName variable
- replaced most of the single line string concatenation with a multiline ... can't change the remainder without breaking the formatting

micchickenburger committed:
Minor stylistic improvements
